src/AaveV2CollectorContractConsolidation.sol

```diff
@@ -0,0 +1,275 @@
+
+        USDC_DECIMALS = USDC.decimals();
+        ETH_USD_ORACLE_DECIMALS = AggregatorV3Interface(ETH_USD_FEED).decimals();
+    }
+
+    /// @notice Lets user pay with USDC for specified token
+    /// @param _token the address of the token to purchase with USDC
+    /// @param _amountOut the amount of token wanted
+    /// @dev User has to approve USDC transfer prior to calling purchase
+    function purchase(address _token, uint256 _amountOut) external {
+        if (_amountOut == 0) revert OnlyNonZeroAmount();
+
+        uint256 amountIn = getAmountIn(_token, _amountOut);
+        uint256 quantity = assets[_token].quantity;
+        uint256 sendAmount = _amountOut == type(uint256).max ? quantity : _amountOut;
+
+        assets[_token].quantity = quantity - sendAmount;
+
+        USDC.transferFrom(msg.sender, AaveV2Ethereum.COLLECTOR, amountIn);
+        ERC20(_token).safeTransferFrom(AaveV2Ethereum.COLLECTOR, msg.sender, sendAmount);
+        emit Purchase(address(USDC), _token, amountIn, sendAmount);
+    }
+
+    /// @notice Returns amount of USDC to be spent to purchase for token
+    /// @param _token the address of the token to purchase
+    /// @param _amountOut the amount of token wanted
+    /// return amountInWithDiscount the amount of USDC used minus premium incentive
+    /// @dev User check this function before calling purchase() to see the amount of USDC required
+    /// @dev User can pass type(uint256).max as amount in order to purchase all
+    function getAmountIn(address _token, uint256 _amountOut) public view returns (uint256 amountIn) {
+        Asset memory asset = assets[_token];
+        if (asset.oracle == address(0)) revert UnsupportedToken();
+
+        if (_amountOut == type(uint256).max) {
+            _amountOut = asset.quantity;
+        } else if (_amountOut > asset.quantity) {
+            revert NotEnoughTokens();
+        }
+
+        uint256 oraclePrice = getOraclePrice(asset.oracle);
+        uint256 exponent = asset.decimals + asset.oracleDecimals - USDC_DECIMALS;
+
+        if (asset.ethFeedOnly) {
+            uint256 ethUsdPrice = getOraclePrice(ETH_USD_FEED);
+            oraclePrice *= ethUsdPrice;
+            exponent += ETH_USD_ORACLE_DECIMALS;
+        }
+
+        // Basis points arbitrage incentive
+        /** 
+            Amount In Calculation with 3% discount (300 basis points)
+            The actual calculation is a collapsed version of this to prevent precision loss:
+            => amountIn = (amountTokenWei / 10^tokenDecimals) * (chainlinkPrice / chainlinkPrecision) * 10^usdcDecimals
+            => amountInWithDiscount = amountIn * 10000 / (10000 + bps)
+            => ie: amountIn = (amountTokenWei / 10^18) * (chainlinkPrice / 10^8) * 10^6
+            =>     amountInWithDiscount = amountIn * 10000 / (10000 + 300) 
+         */
+        amountIn =
+            (((_amountOut * oraclePrice) / 10**exponent) * 10000) / // Amount before discount
+            (10000 + asset.premium);
```

Comment:
How I think this should look like (Let me know what you think):

```
/** 
    The actual calculation is a collapsed version of this to prevent precision loss:
    => amountIn = (amountTokenWei / 10^tokenDecimals) * (chainlinkPrice / chainlinkPrecision) * 10^usdcDecimals
    => amountInWithDiscount = amountIn * 10000 / (10000 + premium)
   
    Example for asset with 18 decimals, chainlink precision of 8 decimals and 3% premium on asset being purchased (300 bps)
    => ie: amountIn = (amountTokenWei / 10^18) * (chainlinkPrice / 10^8) * 10^6
    =>      amountInWithDiscount = amountIn * 10000 / (10000 + 300)  
*/
amountIn = (_amountOut * oraclePrice * 10000) / (10**exponent * (10000 + asset.premium))
```

Reply:
@efecarranza Can you adjust the comment to also look like the above? I think that makes it very clear as to what's happening.